src/paho/mqtt/properties.py

```diff
@@ -0,0 +1,370 @@
+"""
+*******************************************************************
+  Copyright (c) 2017, 2018 IBM Corp.
+
+  All rights reserved. This program and the accompanying materials
+  are made available under the terms of the Eclipse Public License v1.0
+  and Eclipse Distribution License v1.0 which accompany this distribution.
+
+  The Eclipse Public License is available at
+     http://www.eclipse.org/legal/epl-v10.html
+  and the Eclipse Distribution License is available at
+    http://www.eclipse.org/org/documents/edl-v10.php.
+
+  Contributors:
+     Ian Craggs - initial implementation and/or documentation
+*******************************************************************
+"""
+
+import sys
+
+from .packettypes import PacketTypes
+
+class MQTTException(Exception):
+  pass
+
+class MalformedPacket(MQTTException):
+  pass
+
+def writeInt16(length):
+  return bytearray([length // 256, length % 256])
+
+def readInt16(buf):
+  return buf[0]*256 + buf[1]
+
+def writeInt32(length):
+  buffer = [length // 16777216]
+  length %= 16777216
+  buffer += [length // 65536]
+  length %= 65536
+  buffer += [length // 256, length % 256]
+  return bytearray(buffer)
+
+def readInt32(buf):
+  return buf[0]*16777216 + buf[1]*65536 + buf[2]*256 + buf[3]
+
+def writeUTF(data):
+  # data could be a string, or bytes.  If string, encode into bytes with utf-8
+  if sys.version_info[0] < 3:
+      data = bytearray(data)
+  else:
+      data = data if type(data) == type(b"") else bytes(data, "utf-8")
+  return writeInt16(len(data)) + data
+
+def readUTF(buffer, maxlen):
+  if maxlen >= 2:
+    length = readInt16(buffer)
+  else:
+    raise MalformedPacket("Not enough data to read string length")
+  maxlen -= 2
+  if length > maxlen:
+    raise MalformedPacket("Length delimited string too long")
+  buf = buffer[2:2+length].decode("utf-8")
+  zz = buf.find("\x00") # look for null in the UTF string
+  if zz != -1:
+    raise MalformedPacket("[MQTT-1.5.4-2] Null found in UTF data "+buf)
+  for c in range (0xD800, 0xDFFF):
```

Comment:
I fear about the cost of this. We are iterating over a rather large range (2047 item) times the buffer length which could also be rather long.
I would said the if we are iterating over the buffer anyway, it may be cheaper to do:
```
for c in buf:
  if c >= 0xD800 and c <= 0xDFFF:
    raise MalformedPacket
```

Reply:
We could have a 'strict' option in the library which turns such checks on or off?  Not sure what the default would be.

I would have thought the payload would potentially be the biggest piece of data, but yes, the UTF-8 strings could be large too.

Reply:
Fixed as suggested.